### ISSUE TITLE
Unexpected values when using alias that happens to be a (joined) column name

### ISSUE THREAD

**Issue by frankzandhuis** [duckdb/duckdb#13991]
### What happens?

Possibly related to: https://github.com/duckdb/duckdb/issues/13952

When referring to an alias that is also a column name in a joined table, the joined table column is used instead of the alias. See:

```
WITH CTE_years AS (
	SELECT 2022 AS "financialYear", 100 as "myInt", UNION ALL
	SELECT 2023 AS "financialYear", 200 as "myInt", UNION ALL
	SELECT 2024 AS "financialYear", 300 as "myInt",
), CTE_dates AS (
	SELECT '2022-03-01'::date AS "importantDate" UNION ALL
	SELECT '2023-05-01'::date AS "importantDate" UNION ALL
	SELECT '2024-07-01'::date AS "importantDate"
)
SELECT 	t1."importantDate"         AS "importantDate",
	"importantDate".YEAR()     AS "myInt",
	"myInt"+1                  AS "myIntPlusOne",
FROM CTE_dates AS t1
LEFT JOIN CTE_years AS t2
ON t1."importantDate".YEAR() = t2."financialYear";
```

This gives the following output:
```
importantDate	myInt	myIntPlusOne
2022-03-01	2.022	101
2023-05-01	2.023	201
2024-07-01	2.024	301
```

I expect `myIntPlusOne` to be referring my alias `myInt` that I just set (four digit year int). Not the column from the CTE (three digit int).

In this MWE it's clear what's going wrong, but when joining many wide parquet files, I'm unsure how to make use of an alias when it's also a column name in a joined dataset. Even if I pick unused aliases now, in the future the parquets might change and then the aliases are pointing to the incorrect column.

### To Reproduce

I tested this with Dbeaver with DuckDB 1.1.0.

### OS:

Windows 11 Pro, x64

### DuckDB Version:

1.1.0

### DuckDB Client:

JDBC, via Dbeaver

### Full Name:

Frank Zandhuis

### Affiliation:

1 Giant Leap Solutions

### What is the latest build you tested with? If possible, we recommend testing with the latest nightly build.

I have tested with a stable release

### Did you include all relevant data sets for reproducing the issue?

Yes

### Did you include all code required to reproduce the issue?

- [X] Yes, I have

### Did you include all relevant configuration (e.g., CPU architecture, Python version, Linux distribution) to reproduce the issue?

- [X] Yes, I have

**Comment by Mytherin** [comment]
Thanks for the report!

Unfortunately we can't really solve this. When a column exists in the base table, we **have** to refer to that column first, otherwise we break compatibility with all other SQL systems. For example, with a minor rewrite of the `YEAR` function, we can make your example run in Postgres, which then produces the same result as DuckDB:

```sql
WITH CTE_years AS (
      SELECT 2022 AS "financialYear", 100 as "myInt" UNION ALL
      SELECT 2023 AS "financialYear", 200 as "myInt" UNION ALL
      SELECT 2024 AS "financialYear", 300 as "myInt"
  ), CTE_dates AS (
      SELECT '2022-03-01'::date AS "importantDate" UNION ALL
      SELECT '2023-05-01'::date AS "importantDate" UNION ALL
      SELECT '2024-07-01'::date AS "importantDate"
  )
  SELECT     t1."importantDate"         AS "importantDate",
      extract(year from "importantDate")     AS "myInt",
      "myInt"+1                  AS "myIntPlusOne"
  FROM CTE_dates AS t1
  LEFT JOIN CTE_years AS t2
  ON extract(year from t1."importantDate") = t2."financialYear";

 importantDate | myInt | myIntPlusOne 
---------------+-------+--------------
 2022-03-01    |  2022 |          101
 2023-05-01    |  2023 |          201
 2024-07-01    |  2024 |          301
(3 rows)

```

Other SQL systems do not support referring to aliases in the `SELECT` clause, and therefore will always bind to columns in the base table. If we were to prioritize aliases in the `SELECT` clause, we would be producing a different result for the exact same SQL query as other SQL systems, which is undesirable.

**Comment by frankzandhuis** [comment]
Would it be possible to add some sort of DuckDB notation on top of SQL to modify this behavior? 

Perhaps I can explicitly refer to `t1.myInt`, from `myIntPlusOne`, a column which doesn't exist in `t1`. However, then you only need to look if an alias exists outside of the other tables? You are much better in seeing the impact/ downsides of something like this. But it would be really helpful if I could explictly refer to aliases. 

For data processing/ cleansing I am integrating many datasets with similar/ related column names. It would be great to keep a simple naming convention throughout my data processing, without resorting to "unique" column names to ensure correct references.

If the only "correct" solution is to ensure that created aliases are unique, then that might just be it...

**Comment by soerenwolfers** [comment]
I updated an old ticket of mine with a reference to here. My ticket is about column resolution in HAVING clauses instead of SELECT clauses, but there is an interesting connection: While MySQL doesn't allow using aliases from within the same SELECT clause, it does allow using aliases defined in the SELECT clause in a subsequent HAVING clause, and it does give them precedence over the original column names (whereas duckdb sticks to giving preference to the original columns).

**Comment by Mytherin** [comment]
Perhaps we can add a special `alias` table so that we can explicitly refer to aliases in the query? For example:

```sql
select a + 1 as a, alias.a + 1 AS b from (values (42)) t(a);
┌───────┬───────┐
│   a   │   b   │
│ int32 │ int32 │
├───────┼───────┤
│    43 │    44 │
└───────┴───────┘
```

**Comment by soerenwolfers** [comment]
Wouldn't that just move the problem to: "if a table alias or schema named 'alias' exists, we have to pick that instead of the special keyword to keep compatibility with PostgreSQL"?

**Comment by Mytherin** [comment]
Yes, that is true, but all table aliases are explicitly defined in the query itself, whereas columns are defined elsewhere (either in the table definition, or in a CSV/parquet file, etc). As such, when writing a query you are in control of the table aliases, but not of the present columns, which is the main issue here I think.

**Comment by frankzandhuis** [comment]
Sounds like a great idea. Plus, if we make the alias table name a pragma, or other setting, people that are using alias in their naming of tables/schemas/etc could change the name of their alias table. Or we could put this alias approach behind a setting to explicitly enable.

**Comment by soerenwolfers** [comment]
(`self` or `this` could work too)
